include/revng/Pipeline/LibraryRegistry.h

```diff
@@ -0,0 +1,179 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <utility>
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+
+#include "revng/Pipeline/Loader.h"
+
+namespace Pipeline {
+
+/// a library registry is the general way of registering
+/// factories and kinds from down stream libraries.
+///
+/// it is possible that a factory my require to look up
+/// something into the builder and thus this class used to that purpose
+///
+/// for those kinds and containers and pipes that do not require to look inside
+/// the loader look further into this file to fine more declarative classes.
+class LibraryRegistry {
+private:
+  static llvm::SmallVector<LibraryRegistry *, 4> Instances;
+
+public:
+  static void registerAllContainersAndPipes(Loader &Loader) {
+    for (const auto &Reg : Instances)
+      Reg->registerContainersAndPipes(Loader);
+  }
+
+  static void runAllInitializationRoutines() {
+    for (const auto &Reg : Instances)
+      Reg->libraryInitialization();
+  }
+
+  static KindsRegistry registerAllKinds() {
+    KindsRegistry Registry;
+    for (const auto &Reg : Instances)
+      Reg->registerKinds(Registry);
+
+    Kind::init();
+    Granularity::init();
+    return Registry;
+  }
+
+  virtual ~LibraryRegistry() = default;
+
+  virtual void registerContainersAndPipes(Loader &Loader) = 0;
+  virtual void registerKinds(KindsRegistry &KindDictionary) = 0;
+  virtual void libraryInitialization() = 0;
+
+protected:
+  LibraryRegistry() { Instances.push_back(this); }
+};
+
+/// Instantiate a global object of this class for each
+/// kind you whish to register.
+class RegisterKind : LibraryRegistry {
```

Comment:
well, i can, it does nothing toh. there are no base objects, and no base methods, and it does not make any sense to call this objects from anywhere else that the  static methods in the base class.

Reply:
I honestly don't remember the effect of `public` there, I just always put it.
If you think it's appropriate, leave this way `:)`

Reply:
the effect should be that base methods are not visible from the outside unless they are overridden from the current class. And they always are because they are pure.